docs/caching.md

```diff
@@ -0,0 +1,281 @@
+
+## Example 3: When the function body changes
+
+Without stopping and restarting your Streamlit server, let's remove the widget from our app and modify the function's code by adding a `+ 1` to the return value.
+
+```python
+import streamlit as st
+import time
+
+@st.cache(suppress_st_warning=True)
+def expensive_computation(a, b):
+    st.write("Cache miss: expensive_computation(", a, ",", b, ") ran")
+    time.sleep(2)  # This makes the function take 2s to run
+    return a * b + 1  # 👈 Add a +1 at the end here
+
+a = 2
+b = 210
+res = expensive_computation(a, b)
+
+st.write("Result:", res)
+```
+
+The first run is a "Cache miss", but each subsequent run is a cache hit. This is because on first run, Streamlit detected that the function body changed, reran the function, and put the result in the cache.
+
+```eval_rst
+.. tip::
+   if you change the function back the result will already be in the Streamlit cache from a previous run. Try it out!
+```
+
+## Example 4: When an inner function changes
+
+Let's make our cached function depend on another function internally:
+
+```python
+import streamlit as st
+import time
+
+def inner_func(a, b):
+    st.write("inner_func(", a, ",", b, ") ran")
+    return a * b
+
+@st.cache(suppress_st_warning=True)
+def expensive_computation(a, b):
+    st.write("Cache miss: expensive_computation(", a, ",", b, ") ran")
+    time.sleep(2)  # This makes the function take 2s to run
+    return inner_func(a, b) + 1
+
+a = 2
+b = 210
+res = expensive_computation(a, b)
+
+st.write("Result:", res)
+```
+
+What you see is the usual:
+The first run results in a cache miss.
+Every subsequent rerun results in a cache hit.
+
+But now let's try modifying the `inner_func()`:
+
+```python
+import streamlit as st
+import time
+
+def inner_func(a, b):
+    st.write("inner_func(", a, ",", b, ") ran")
+    return a ** b  # 👈 Change the * to ** here
+
+@st.cache(suppress_st_warning=True)
+def expensive_computation(a, b):
+    st.write("Cache miss: expensive_computation(", a, ",", b, ") ran")
+    time.sleep(2)  # This makes the function take 2s to run
+    return inner_func(a, b) + 1
+
+a = 2
+b = 21
+res = expensive_computation(a, b)
+
+st.write("Result:", res)
+```
+
+Even though `inner_func()` is not annotated with [`@st.cache`](api.html#streamlit.cache), when we edit its body we cause a "Cache miss" in the outer `expensive_computation()`.
+
+That's because Streamlit always traverses your code and its dependencies to verify that the cached values are still valid. This means that while developing your app you can edit your code freely without worrying about the cache. Any change you make to your app, Streamlit should do the right thing!
+
+Streamlit is also smart enough to only traverse dependencies that belong to your app, and skip over any dependency that comes from an installed Python library.
+
+
+## Example 5: Use the global cache to speed up your app for all users
```

Comment:
When you say "global cache" it makes it seem that we have two types of cache: global and local. So just say "cache" here, for clarity.